### PR TITLE
Dt workingfrommaster

### PR DESCRIPTION
**Pull request by JYoung413**
Drivetrain rewritten:
- More idiomatic WPILib
- Increased efficiency through using WPILIB RobotDrive methods opposed to where we made our own
- GyroWrapper class for PID controllers
- Deleted unused variable from shooter class
- Mutatis Mutandis

### CODE REVIEW

File: src/lib/GyroWrapper.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package lib;
+
+import edu.wpi.first.wpilibj.AnalogGyro;
+import edu.wpi.first.wpilibj.PIDSource;
+import edu.wpi.first.wpilibj.PIDSourceType;
+
+public class GyroWrapper implements PIDSource {
+	private PIDSourceType pidSource;
+	private AnalogGyro gyro;
+	public GyroWrapper() {
```
Comment (MiloDavis) [inline src/lib/GyroWrapper.java]: Unless this constructor is specifically required by the library, delete it.

File: src/lib/GyroWrapper.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		//empty
+	}
+	/**
+	 * Creates a EncoderWrapper PIDSource
+	 * @param pidSource The pidSource type
+	 * @param enc The encoder used for the PIDSource data
+	 * @param encSourceA the first digital port for the encoder
+	 * @param encSourceB the second digital port for the encoder
+	 */
+	public GyroWrapper(PIDSourceType pidSource,AnalogGyro gyro,int gyroSource) {
```
Comment (MiloDavis) [inline src/lib/GyroWrapper.java]: You need spaces after commas

File: src/com/nutrons/nu17/RobotMap.java [hunk 3]
```diff
@@ -3,10 +3,10 @@
 public class RobotMap {
 
 	// Ports of wheels
-	public static final int FRONT_LEFT = 0;
-	public static final int FRONT_RIGHT = 1;
-	public static final int BACK_LEFT = 2;
-	public static final int BACK_RIGHT = 3;
+	public static final int LEFT_DRIVE_MOTOR_A = 0;
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/RobotMap.java]: Giving the direction "front, center, back" is clearer when translating this into wiring.

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 4]
```diff
@@ -1,104 +1,115 @@
...
 import edu.wpi.first.wpilibj.PIDSourceType;
+import edu.wpi.first.wpilibj.RobotDrive;
 import edu.wpi.first.wpilibj.Talon;
 import edu.wpi.first.wpilibj.command.Subsystem;
+import lib.EncoderWrapper;
+import lib.GyroWrapper;
+import lib.HoldPID;
 
+/**
+ * Drivetrain code four a 4 wheel drive
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java]: We know this is code, so you don't need to say it.

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 5]
```diff
@@ -1,104 +1,115 @@
 package com.nutrons.nu17.subsystems;
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java:1]: Check the formatting on this entire class.  Comments should end on the line before the class starts.  Indentation should be consistent.  Fix this stuff in this PR

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 6]
```diff
@@ -1,104 +1,115 @@
...
+	
+	// Motors
+		public Talon leftDriveA = new Talon(RobotMap.LEFT_DRIVE_MOTOR_A);
+		public Talon leftDriveB = new Talon(RobotMap.LEFT_DRIVE_MOTOR_B);
+		public Talon rightDriveA = new Talon(RobotMap.RIGHT_DRIVE_MOTOR_A);
+		public Talon rightDriveB = new Talon(RobotMap.RIGHT_DRIVE_MOTOR_B);
+    
+	// Sensors
+		private final AnalogGyro GYRO = new AnalogGyro(RobotMap.GYRO);
+		// Encoder Ports Not FINAL Needs Tuning!
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java]: This comment should read:
`// TODO: Tune PID constants`

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 7]
```diff
@@ -1,104 +1,115 @@
...
 		//empty
 	}
+	
+	// Motors
+		public Talon leftDriveA = new Talon(RobotMap.LEFT_DRIVE_MOTOR_A);
+		public Talon leftDriveB = new Talon(RobotMap.LEFT_DRIVE_MOTOR_B);
+		public Talon rightDriveA = new Talon(RobotMap.RIGHT_DRIVE_MOTOR_A);
+		public Talon rightDriveB = new Talon(RobotMap.RIGHT_DRIVE_MOTOR_B);
+    
+	// Sensors
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java:30]: Fix the indentation here.  The declarations should be indented to the same level as the comment.

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 8]
```diff
@@ -1,104 +1,115 @@
...
-
-		@Override
-		public PIDSourceType getPIDSourceType() {
-			return null;
+		
+		/**
+		 * Resets the encoder.
+		 */
+		public void resetEncoder() {
+			this.ENC_A.reset();
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java]: Extra newlines

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 9]
```diff
@@ -1,104 +1,115 @@
...
-		// Gets the speed at which to run the wheels at and uses it to drive the robot
-		@Override
-		public void pidWrite(double output) {
-			drive(output, -output, output, -output);
+		/**
+		 * Resets the gyro.
+		 */
+		public void resetGyro() {
+			this.GYRO.reset();
+
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java]: Extra newline

File: src/com/nutrons/nu17/subsystems/TwinShooter.java [hunk 10]
```diff
@@ -59,8 +59,7 @@ public TwinShooter() {
 			encWrap2, 
 			twinHold2);
 	
-	private static double holdShootA;
-	private static double holdShootB;
+
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/TwinShooter.java:62]: Too many newlines

File: src/lib/GyroWrapper.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+		setPIDSourceType(pidSource);
+		gyro = new AnalogGyro(gyroSource);
+	}
+	@Override
```
Comment (MiloDavis) [inline src/lib/GyroWrapper.java]: The javadoc goes before the `@Override` annotation.

File: src/lib/GyroWrapper.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	public void setPIDSourceType(PIDSourceType pidSourceType) {
+		pidSource = pidSourceType;
+	}
+	@Override
+	/** 
+	 * Returns the PIDSourceType
+	 */
+	public PIDSourceType getPIDSourceType() {
+		return pidSource;
+	}
```
Comment (MiloDavis) [inline src/lib/GyroWrapper.java:36]: Need a line break between methods

**Review by MiloDavis — COMMENTED** [review]
Decent, but watch your style and logic.

**Comment by JYoung413** [comment]
Fixed issues, please review

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 13]
```diff
@@ -14,102 +14,97 @@
...
-		
-
-    // Drive
-		public RobotDrive drive = new RobotDrive(leftDriveA, leftDriveB,
-				rightDriveA, rightDriveB);
-		
+	private final AnalogGyro GYRO = new AnalogGyro(RobotMap.GYRO);
+
+	// TODO: Tune Ports
+	private final Encoder ENC_A = new Encoder(RobotMap.DT_ENCODER_1, RobotMap.DT_ENCODER_2);
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java]: Make naming consistent with wheels and motors

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 14]
```diff
@@ -14,102 +14,97 @@
...
 	}
-	
+
 	// Motors
-		public Talon leftDriveA = new Talon(RobotMap.LEFT_DRIVE_MOTOR_A);
-		public Talon leftDriveB = new Talon(RobotMap.LEFT_DRIVE_MOTOR_B);
-		public Talon rightDriveA = new Talon(RobotMap.RIGHT_DRIVE_MOTOR_A);
-		public Talon rightDriveB = new Talon(RobotMap.RIGHT_DRIVE_MOTOR_B);
-    
+	public Talon leftDriveA = new Talon(RobotMap.FRONT_LEFT);
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java]: Same naming convention here too

File: src/com/nutrons/nu17/subsystems/Drivetrain.java [hunk 15]
```diff
@@ -14,102 +14,97 @@
...
+	private EncoderWrapper encWrap = new EncoderWrapper(PIDSourceType.kDisplacement,
+			ENC_A,
+			RobotMap.DT_ENCODER_1,
+			RobotMap.DT_ENCODER_2);
+	private GyroWrapper gyroWrap = new GyroWrapper(PIDSourceType.kDisplacement,
+			GYRO,
+			RobotMap.GYRO);
+
+	// Holders
+	private HoldPID driveHoldA = new HoldPID();
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/subsystems/Drivetrain.java]: These need real names.  A B and C are not descriptive enough

**Review by MiloDavis — COMMENTED** [review]
The biggest issue is variable naming.  Fix and I'll merge

File: src/com/nutrons/nu17/RobotMap.java [hunk 16]
```diff
@@ -36,8 +36,8 @@
 	public static final int MICRO_SWITCH = 0;
 
 	// Ports of encoders
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/RobotMap.java:38]: If this is attached to a wheel, tell me which wheel

**Review by MiloDavis — CHANGES_REQUESTED** [review]

File: src/com/nutrons/nu17/RobotMap.java [hunk 17]
```diff
@@ -9,12 +9,12 @@
 	public static final int BACK_RIGHT = 3;
 	
 	// Ports of intake
-	public static final int ROLLER_A = 0;
-	public static final int ROLLER_B = 0;
+	public static final int ROLLER_FIRST = 0;
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/RobotMap.java]: This is just as bad.  I should be able to read the variable and point to the bit of the robot it attaches to.

File: src/com/nutrons/nu17/RobotMap.java [hunk 18]
```diff
@@ -36,18 +36,18 @@
 	public static final int MICRO_SWITCH = 0;
 
 	// Ports of encoders
-	public static final int DRIVE_DISTANCE_ENCODER_PORT_1 = 1;
-	public static final int DRIVE_DISTANCE_ENCODER_PORT_2 = 2;
+	public static final int DRIVE_DISTANCE_ENCODER_1 = 1;
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/RobotMap.java]: 1 and 2 are not helpful Josh.  Once again, I should be able to look at the variable and find it on the Robot.  1 and 2 have no meaning on the robot.

File: src/com/nutrons/nu17/RobotMap.java [hunk 19]
```diff
@@ -36,18 +36,18 @@
...
 	public static final int FLY_ENCODER_BACK_RIGHT = 6;
 	public static final int FLY_ENCODER_FRONT_LEFT = 7;
 	public static final int FLY_ENCODER_BACK_LEFT = 8;
-	public static final int TWIN_A_ENCODER_PORT_1 = 9;
-	public static final int TWIN_A_ENCODER_PORT_2 = 10;
-	public static final int TWIN_B_ENCODER_PORT_1 = 11;
-	public static final int TWIN_B_ENCODER_PORT_2 = 12;
-	public static final int SHOOT_ENCODER_PORT_1 = 0;
-	public static final int SHOOT_ENCODER_PORT_2 = 1;
+	public static final int TWIN_LEFT_ENCODER_1 = 9;
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/RobotMap.java:47]: What do 1 and 2 mean.  Front and back?  Left and right?  Make this explicit.

**Review by MiloDavis — CHANGES_REQUESTED** [review]

File: src/com/nutrons/nu17/RobotMap.java [hunk 20]
```diff
@@ -7,10 +7,10 @@
 	public static final int	BACK_LEFT = 1;
 	public static final int FRONT_RIGHT = 2;
 	public static final int BACK_RIGHT = 3;
-	
+	// Paramter naming convention : Subsystem SIDE_PLACEMENT
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/RobotMap.java:10]: Put this at the top of the file

File: src/com/nutrons/nu17/RobotMap.java [hunk 21]
```diff
@@ -7,10 +7,10 @@
...
 	// Ports of intake
-	public static final int ROLLER_FIRST = 0;
-	public static final int ROLLER_SECOND = 0;
+	public static final int ROLLER_FRONT = 0;
```
Comment (MiloDavis) [inline src/com/nutrons/nu17/RobotMap.java:12]: Missing subsystem

File: .classpath [hunk 22]
```diff
@@ -6,5 +6,6 @@
 	<classpathentry kind="var" path="opencv" sourcepath="opencv.sources"/>
 	<classpathentry kind="var" path="cscore" sourcepath="cscore.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="var" path="USERLIBS_DIR/CTRLib.jar"/>
```
Comment (MiloDavis) [inline .classpath:9]: Is this library going to be on everyone's system

**Review by MiloDavis — CHANGES_REQUESTED** [review]

File: .classpath (hunk 22 above)
Comment (MiloDavis) [reply .classpath:9]: @JYoung413

**Review by MiloDavis — APPROVED** [review]

**Pull request merged**